src/descriptor/create_descriptor.rs

```diff
@@ -0,0 +1,446 @@
+//! Tools for creating Descriptor and witness stack from given scriptpubkey and corresponding
+//! scriptsig and witness.
+//!
+//! Figures out the the type of descriptor based on scriptpubkey, witness and scriptsig.
+//! Outputs a descriptor and witness_stack for all descriptors which can be directly fed into the
+//! interpreter. All script_sig and witness are translated into a single witness stack
+//! Vec<StackElement>
+//! 1) Pkh: translates scriptsig to witness_stack<StackElement>, validates sig and pubkey hash
+//! and outputs Pkh descriptor
+//! 2) Wphk: translates witness to witness_stack<StackElement>, validates sig and pubkey hash
+//! and outputs a Wpkh descriptor
+//! 3) Wsh: pops witness script and checks wsh output hash, translates remaining witness elements
+//! to Vec<StackElement> and outputs a Wsh descriptor. Does not check miniscript inside the descriptor
+//! 4) Bare: translates script_sig to witness_stack<StackElement> and script_pubkey to miniscript
+//! 5) Sh: Checks redeem_script hash, translates remaining elements from script_sig to
+//! witness_stack<StackElement> and redeem script to miniscript. Does not check the miniscript
+//! 6) ShWpkh: Sh: Checks redeem_script hash, translates remaining elements from script_sig to
+//! witness_stack<StackElement> and validates Wpkh sig, pubkey.
+//! 7) ShWsh: Checks witness script hash, pops witness script and converts it to miniscript.
+//! translates the remaining witness to witness_stack<StackElement>
```

Comment:
1. I think this whole block comment should actually go on `pub fn witness_stack` rather than on the module.
2. I don't think we should be verifying signatures in this function; that seems like a responsibility of the interpreter (and it seems like you're already doing it there)

Reply:
The signatures are being checked for non-miniscript `Pkh/WPkh/ShWpkh` descriptors. 

For `Pkh/WPkh/ShWpkh` types, I should only check the pubkey hash matches and not the signatures. And the interpreter should then check only the signatures?

Reply:
I think so, yeah. Reason is that I want to be able to abuse `witness_stack` by giving it only the witnessScript/redeemScript and no actual satisfying data, and have it give me the `Descriptor` and an empty `Vec<StackElement>` back.

This gives me the ability to get a descriptor from a scriptpubkey/scriptsig/witness script.

Reply:
So, `Pkh/WPkh/ShWpkh` should return empty `Vec<StackElement>` or vec![sig pk] or vec![sig] since is in descriptor.

Reply:
Sorry, understood what you meant finally with the above comment. 

For `Pkh` if only pk is supplied as `witness` without a sig. It will return empty. If it provided with a `sig`, that will return `[StackElement::Witness(sig)]`. A similar logic for `Wpkh/Shwpkh`

Reply:
Yep exactly :)